### ISSUE TITLE
cargo publish --all

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/cargo#1169]
It would be nice to have a flag to `cargo publish` which publishes all local packages in a DAG fashion.

**Comment by wycats** [comment]
This would be awesome ;)

**Comment by huonw** [comment]
I suspect we may want to have a story with https://github.com/rust-lang/cargo/issues/883 if/when we implement this: I could easily imagine unintentionally publishing crates by not realising that I'm depending on them.

**Comment by Botev** [comment]
Is there any progress on this? I think this would a very good thing to have. One of the nice things about crates is that they are modular, thus reduce the compile time of larger projects. However, if you are developing a large library compile times can get pretty slow (when it hits >2min I start to compile after 50% of the fixing, so it can compile while I'm fixing the other 50% of the issues). In this case you can split the library into sub-crates, where each sub-crate is in some sense standalone, or a build-up over "core" structure. However, each one of them would not make any sense on its own. Publishing each individually does not make any sense also and not packaging it together. If we can have this kind of publish --all this would give as an option to reduce compile times without any drawback from the user and publishing side of the library.

**Comment by matthiasbeyer** [comment]
Any updates whether this will be implemented and when?

**Comment by wycats** [comment]
@matthiasbeyer I think this fell through the cracks when we did [workspaces](https://github.com/rust-lang/cargo/issues/2878). This issue was supposed to track wrapping `--all` up for `publish` but I guess it got lost since it was originally opened earlier?

@alexcrichton what's the status?

**Comment by alexcrichton** [comment]
@wycats AFAIK this has always been in the 'nice to have' category and hasn't progressed to the 'someone has put time into designing this' category.

**Comment by matthiasbeyer** [comment]
One thing I just thought about: When doing a workspace release, cargo should build everything and after everything is fine, publish all crates at once... if that is even possible. Not like "build & publish the first crate and continue for each crate" but rather "build everything, then publish everything".

**Comment by kpcyrd** [comment]
It seems that in a `foo-rs`, `foo-rs/foo-sys` project layout, you can't `cargo package` in `foo-rs` before `foo-rs/foo-sys` has been published.

**Comment by Nemo157** [comment]
Something else I would want that I don’t think that can handle is only publishing updated packages, if one of the packages already exists at its current version number it should be downloaded and verified that the new package is identical.

**Comment by softprops** [comment]
being new to workspaces but not to cargo this feels very much like a paper cut. My finger just git nipped when I tried to publish a new workspace project by following the docs

![screen shot 2018-10-09 at 4 45 01 pm](https://user-images.githubusercontent.com/2242/46699182-036a9280-cbe7-11e8-9b1b-b4882df24cec.png)

...then realized that order matters. The validation of packages will fail of one of the workspace packages depends on another in the same release but which may not have been published first. In my case it's a very simple ordering but for those new to cargo, something like `cargo publish --all` could get the ordering right without putting a burden on me at all. Something like that would be a much nicer new user experience, also a much nicer convenience to those that have learned the ordering semantics for the more more manual publishing approach.

**Comment by matthiasbeyer** [comment]
It would definitively help me with [imag](https://imag-pim.org) where I [publish over 50 crates in one release](http://git.imag-pim.org/imag/tree/scripts/release.sh)!

**Comment by torkleyy** [comment]
This sounds useful, but I'm wondering what the exact behaviour should be. What if you have a `path` dependency and there's no version specified, should Cargo modify the `Cargo.toml` for that?

**Comment by mitchmindtree** [comment]
@torkleyy interesting point. It probably should not modify the Cargo.toml at all, but it would be worth considering whether or not cargo should allow for `path` dependencies but only for crates within the same workspace.

E.g. If building `foo` and a dependency on `bar` is specified via a path, perhaps cargo could implicitly publish the new version of `foo` with the version of `bar` that was published during the same `cargo publish --all`? This would also allow for cargo to easily build everything first locally before publishing any of the packages. If for some reason a crate within the workspace would depend

### Simplified Steps

I'm imagining `cargo publish --all` should do something like the following:

1. Check manifests and version validity of all packages before starting to build any.
2. Create DAG of workspace packages and determine the build order.
3. Build packages. If any failures occur, bail out of the whole process.
4. If all manifests and valid and packages are built, publish all packages in the order in which they were built.

It would be worth considerinng if step 4 should be a special "atomic step" recognised by crates.io so that if for some reason the net drops out or there's a crash the user doesn't end up with only half of their packages published.

**Comment by Nemo157** [comment]
@mitchmindtree I would like to also see an extra step between 3 and 4 doing full workspace package validation to replace the current pre-publish validation. This does the normal per-package validation steps with 2 changes:

 1. crates.io is queried for an exact version match of any of the crates. If there is an exact version match then the contents are compared to make sure the crate hasn't changed, if they match the crate is removed from the set to publish. (Basically making publishing a crate idempotent so that you can republish a workspace where you have only updated some of the crates.)

 2. during this validation crates can depend on crates that are either available on crates.io or are part of the current set to publish.

**Comment by torkleyy** [comment]
I started working on this here: https://gitlab.com/torkleyy/cargo-publish-all

**Comment by sehz** [comment]
https://gitlab.com/torkleyy/cargo-publish-all works well.

**Comment by softprops** [comment]
Thanks for sharing. This looks like a good stop gap for not having this as a built in. The main reason I'd still champion this as a built-in is for consistency as other built-in commands support the --all flag.

**Comment by KSXGitHub** [comment]
What's the status of this issue?

**Comment by pksunkara** [comment]
Sorry for the drive by, but in case anyone is wondering, I got inspired by [lerna](https://lerna.js.org) and made https://github.com/pksunkara/cargo-workspaces. rust-analyzer and chalk are using it to auto publish weekly.

**Comment by leighmcculloch** [comment]
> * `--dry-run` should work correctly (will need to pretend that the previous crates have been published, maybe via `patch`?)

I am currently using the following process to simulate dry-run publishing a workspace of crates. The process builds a `vendor/` directory containing all the published dependencies, and all the crates in the workspace.

Potentially cargo could use this same, or similar, process.

The process is:

- Temporarily remove git dependencies from workspace Cargo.toml.
- Vendor all dependencies in a `vendor/` directory.
- Undo changes to Cargo.toml.
- Package all crates, verification disabled.
- Extract the *.crate files into `vendor/`.
- Add a `.cargo-checksum.json` file to each crate.
- Package all crates, verification enabled using `vendor/` as source replacement for `crates.io`.

This process exists in the GitHub Actions workflow below and I use it on ~6 repositories successfully to perform verification on workspaces before starting publishing.

https://github.com/stellar/actions/blob/70940b15e/.github/workflows/rust-publish-dry-run.yml

**Comment by alamb** [comment]
I would like to say thank you to @epage  and everyone who worked on this feature. It has made my life publishing crates from https://github.com/apache/arrow-rs much better

**Comment by brainstorm** [comment]
> I started working on this here: https://gitlab.com/torkleyy/cargo-publish-all

@torkleyy I just tested your tool on my workspace I'm trying to publish on https://github.com/umccr/htsget-rs/tree/better_ci but no output is returned and fails with exit code 1.

OTOH, trying to publish my first set of crates ever is turning to be a fun circular ride... not the UX I had on the rest of cargo subcommands :_S

/cc @mmalenic

```shell
% cargo publish -p htsget-http-actix
    Updating crates.io index
   Packaging htsget-http-actix v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-http-actix)
error: failed to prepare local package for uploading

Caused by:
  no matching package named `htsget-http-core` found
  location searched: registry `crates-io`
  required by package `htsget-http-actix v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-http-actix)`

% cargo publish -p htsget-http-core
    Updating crates.io index
   Packaging htsget-http-core v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-http-core)
   Verifying htsget-http-core v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-http-core)
error: failed to verify package tarball

Caused by:
  no matching package named `htsget-search` found
  location searched: registry `crates-io`
  required by package `htsget-http-core v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/target/package/htsget-http-core-0.1.0)`

% cargo publish -p htsget-search
    Updating crates.io index
   Packaging htsget-search v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-search)
   Verifying htsget-search v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-search)
error: failed to verify package tarball

Caused by:
  no matching package named `htsget-test-utils` found
  location searched: registry `crates-io`
  required by package `htsget-search v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/target/package/htsget-search-0.1.0)`

% cargo publish -p htsget-test-utils
   Updating crates.io index
   Packaging htsget-test-utils v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-test-utils)
   Verifying htsget-test-utils v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/htsget-test-utils)
error: failed to verify package tarball

Caused by:
  no matching package named `htsget-http-core` found
  location searched: registry `crates-io`
  required by package `htsget-test-utils v0.1.0 (/Users/rvalls/dev/umccr/htsget-rs/target/package/htsget-test-utils-0.1.0)`
```

**Comment by epage** [comment]
@brainstorm Looks like you have a cycle involving a dev-dependency.  The simplest way of resolving that is to remove the cycle by removing the `version` field in `dev-depencency`s that also have a `path`.  `cargo add` does this by default.  The main downside is that people running tests from the registry won't be able to.  In practice, this is just crater (the tool that tests rust changes across all of crates.io).

In `cargo-release`, we just expect people to use the workaround above.  I **think** `cargo-smart-release` has some special logic to handle cycles.

**Comment by epage** [comment]
Trying to summarize this thread with some of my thoughts

Prior art
- https://gitlab.com/torkleyy/cargo-publish-all
- https://github.com/pksunkara/cargo-workspaces
- https://github.com/crate-ci/cargo-release

High-level path
- [ ] #11062
- [ ] Support multi-package packaging
- [ ] Support multi-package publishing
- [ ] Atomic multi-package publish via crates-io-cargo-teams#82  (likely not a blocker for this issue, can be spun out into its own)

This intentionally leaves of "don't publish if its already published" as I see that as separate, though related, to this issue and it has logic/policies to be worked out.

## Multi-package packaging

Requirements
- Add to `cargo package` the normal package selection CLI (`--package`, `--exclude`, `--workspace`), including behavior (default-members)
- Dry-run can verify all of the packages

For dry-run, we'd need to build in-order and patch in the .crates that we already packaged (see https://github.com/rust-lang/cargo/issues/1169#issuecomment-1254260875).

We likely should detect dependency cycles to give people errors early (https://github.com/rust-lang/cargo/issues/1169#issuecomment-1372361912).

## Multi-package publishing

Because this isn't atomic, we should try to do all verification upfront so there aren't errors along the way
- Package everything (including verification) before any publishes
  - Also important once we have atomic publish as we'll need all of the `.crate`s
- Checking if any of the proposed versions are already published
- Ensuring dependencies are published or will be published as part of this
- ...

When publishing, we'll also have to do it in-order. 

If a publish PUT fails, we need to be clear about what didn't get published for people to recover.

We should track the wait-for-publish timeouts from #11602 on a per-package basis so
- Publishing a package only waits for its dependencies to be available (or timeout)
- At the tail end, we only have one timeout if we are waiting on multiple leaf packages

## Additional Info

For me, the biggest open question is how to build the DAG for packaging and then pass that up to publishing.  Last I looked, the main ways I saw for interacting with the DAG was compilation which is too heavy handed for what we need.

For me, the biggest area of complexity is the dependency + timeout tracking while publishing in-order.

**Comment by Nemo157** [comment]
> We likely should detect dependency cycles to give people errors early (https://github.com/rust-lang/cargo/issues/1169#issuecomment-1372361912).

Dependency cycles are actually allowed, as long as the package names have been previously registered on crates.io without cycles. There's no version checking performed by crates.io, just a simple name check, so you can publish crates that depend on future unpublished versions of other crates.

**Comment by epage** [comment]
I've edited it to clarify that is for when verifying as that does not support cycles (minus the automatically removed dev-dependencies)

**Comment by epage** [comment]
An interesting challenge for us to keep in mind with this is registries that rate-limit, both in terms of finding the right strategy for backing off but also the right UX so someone doesn't [publish 300 crates](https://rust-lang.zulipchat.com/#narrow/stream/318791-t-crates-io/topic/Throttling.20issues) and it takes 24 hours without any clear indication.

**Comment by torhovland** [comment]
This issue is `S-accepted`, but it seems to depend on #10948, which isn't. Any chance that can get accepted?

**Comment by torhovland** [comment]
Now that #13947 is merged, it's time to look into multi-package publishing:

> * Package everything (including verification) before any publishes
> * Checking if any of the proposed versions are already published
> * Ensuring dependencies are published or will be published as part of this
> * When publishing, we'll also have to do it in-order.

What's a good way to tackle this:

> the biggest open question is how to build the DAG for packaging and then pass that up to publishing.

Would it make sense for `cargo package` to output the packaged order to a file?

**Comment by epage** [comment]
- Package everything (including verification) before any publishes (mostly done via #13947)
- Checking if any of the proposed versions are already published (#14338)
- Ensuring dependencies are published or will be published as part of this
- When publishing, we'll also have to do it in-order.

> Would it make sense for packaging to output the packaged order to a file, and let publishing read that? Or should publishing simply re-run the same ordering logic that packaging did, almost like a dry-run packaging?

imo we shouldn't do sideband communication (using the filesystem to pass state from one function to another).

If we could have a single graph for both, that would be ideal as it removes the risk of the two graph traversals disagreeing.

I think ideally, we'd adjust the packaging abstraction so `cargo package` and `cargo publish` call into the same graph generation (and registry inferring() code and pass that to a graph packaging function and then publish walks the graph to publish everything.

**Comment by epage** [comment]
Was hoping to do a call for testing but ran into #14721 which breaks some ways of testing this feature.

**Comment by foresterre** [comment]
Are there ways of testing which can already help?

**Comment by epage** [comment]
Feel free to give it a try in nightly.  You don't need more formalized testing instructions to give it a try.

**Comment by epage** [comment]
Tried to test it again and ran into another issues, see #14789

**Comment by nyurik** [comment]
I ran into a related issue: my workspace has 3 packages that are chain-dependent (varnish -> macros -> sys). The version is the same for all of them, and gets updated at the same time. All 3 packages' `Cargo.toml` use `version.workspace = true`, and in their dependencies use the needed `varnish-*.workspace = true` . The workspace one looks like this:

```toml
[workspace.package]
version = "0.2.2"

[workspace.dependencies]
varnish = { path = "./varnish", version = "0.2.2" }
varnish-macros = { path = "./varnish-macros", version = "0.2.2" }
varnish-sys = { path = "./varnish-sys", version = "0.2.2" }
```

`cargo publish --dry-run` fails due to `failed to select a version for the requirement` because the sub-package hasn't been published yet -- because 0.2.2 has not been published yet.

**Comment by epage** [comment]
### Call for testing

`cargo publish --workspace` enables publishing multiple packages in a workspace in a single cargo command invocation.

We're looking for feedback to see if this can be stabilized!

#### Basic steps:

Requirements:
- 1.85.0-nightly (2024-12-01) or newer

Try `cargo publish -Zpackage-workspace` in your existing workflows and with various combinations of multiple packages being selected (`--workspace`, `--exclude`, `-p foo -p bar`)

Support in third-party release tools:
- `cargo release` v0.25.14 as `cargo +nightly release -Zworkspace-publish`

In particular, consider
- Publish order is correct?
- Handling of cycles?
- Correctly waiting between published packages?
- `--dry-run` works?

Note:
- You are still subject to crates.io rate limits and publish may fail in the middle of a workspace, see #13530

**Please leave feedback on this issue**

**Comment by marc2332** [comment]
hey @epage ! Just tried it to release [freya v0.3.0-rc.0](https://crates.io/crates/freya/0.3.0-rc.0) and it seems to have worked perfectly!


- [x] Publish order is correct
- (I don't know what cycles are) Handling of cycles
- [x] Correctly waiting between published packages?
- [x] --dry-run works

Here is the command I ran (with and without `--dry-run`):

```sh
cargo +nightly publish --features skia-engine --workspace -Z package-workspace --exclude "examples" --exclude "freya-installer-example"
```

Love this, I hope it gets stabilized soon! ✨ 🦀

**Comment by kpreid** [comment]
I just tried out the feature with `cargo 1.85.0-nightly (d73d2caf9 2024-12-31)`. All the feedback I can think of:

* Once I got it correctly invoked, it worked fine.
* I really appreciate that the verification step now verifies all packages before any of them are published. That’s the one thing I could not do just by scripting multiple `cargo publish` commands.
* It was an unpleasant surprise that `cargo publish --workspace` does not automatically skip `package.publish = false` packages (like my `xtask` and heavy integration tests), but aborts instead. I can see why doing so automatically might be worse (e.g. someone _forgot_ to write `package.publish = false`), but I think there should be an option which means “exclude all `publish = false`”, for those workspaces where that is the correct choice. (Or, more foolproof, require every package to specify *some* `publish` value first?)
    * (`workspace.default-members` is not an appropriate solution here because that would stop tests in the unpublished packages from running by default.)
* Cargo printed “You may press ctrl-c to skip waiting; the crate should be available shortly.” as it usually does. I wonder what this does to publishing multiple packages — surely it must either cancel the remaining packages, run the risk of erroring out due to not waiting before publishing the next package, or do nothing. Perhaps this message should not be printed in this case. Perhaps a different message should be printed while it waits (or at other times too), indicating the number of packages remaining to be published.
* I worry about the situation where uploading fails halfway through. Would I have to specify exactly the packages not already published when trying again, or would cargo be able to DWIM and publish only the missing ones?
* During the packaging step, it printed `Updating crates.io index` for every package. This seems like unnecessary extra traffic to crates.io; a drop in the bucket for crates.io itself but it could be bad for a user on a high-latency network trying to publish many packages.
* The call for testing has the wrong feature name (`-Zpublish-workspace`, not `-Zpublic-dependency`).

**Comment by weihanglo** [comment]
> does not automatically skip `package.publish = false` packages (like my `xtask` and heavy integration tests)

IMO this is a bug. You cannot even publish a single crate with `publish = false`.


> I worry about the situation where uploading fails halfway through. Would I have to specify exactly the packages not already published when trying again, or would cargo be able to DWIM and publish only the missing ones?

<https://github.com/rust-lang/cargo/issues/13397> might be related

> During the packaging step, it printed Updating crates.io index for every package. This seems like unnecessary extra traffic to crates.io

This is likely because Caarg were building a lockfile for each package. I am aware of it, and think it should be fixed. Currently specifying `--offline` helps (of course you need to fetch all dependencies upfront).

**Comment by marc2332** [comment]
I was surprised as well that it didn't skip those crates with `publish = false`, hence why I used `--exclude` in my command.

**Comment by epage** [comment]
@kpreid 

> It was an unpleasant surprise that cargo publish --workspace does not automatically skip package.publish = false packages (like my xtask and heavy integration tests), but aborts instead. I can see why doing so automatically might be worse (e.g. someone forgot to write package.publish = false), but I think there should be an option which means “exclude all publish = false”, for those workspaces where that is the correct choice. (Or, more foolproof, require every package to specify some publish value first?) 

@weihanglo 

> IMO this is a bug. You cannot even publish a single crate with publish = false.

`cargo publish -p publish-disabled` will error out, rather than skip the package.  If we automatically skip packages, we'd need to decide in what circumstances we skip vs error *and why*.

**Comment by epage** [comment]
@kpreid 

> Cargo printed “You may press ctrl-c to skip waiting; the crate should be available shortly.” as it usually does. I wonder what this does to publishing multiple packages — surely it must either cancel the remaining packages, run the risk of erroring out due to not waiting before publishing the next package, or do nothing. Perhaps this message should not be printed in this case. Perhaps a different message should be printed while it waits (or at other times too), indicating the number of packages remaining to be published.

Might be good to open an issue for this.

**Comment by epage** [comment]
> > During the packaging step, it printed Updating crates.io index for every package. This seems like unnecessary extra traffic to crates.io
> 
> This is likely because Caarg were building a lockfile for each package. I am aware of it, and think it should be fixed. Currently specifying `--offline` helps (of course you need to fetch all dependencies upfront).

This comes about from an implementation detail.  We are effectively stringing together separate commands (`package`, `build`).  Each command loads the registry's index separately.  Each will see that it hasn't yet checked if there is a newer version of the index available and will check.

We could fake it by overriding the config to be in offline mode but we don't have an easy way to layer that on top of the existing config, so we'd be mutating the config and, ideally, trying to roll back without being able to use RAII because we'd need to have a `&mut` while we share that with other commands.

I thought we had an issue about checking the registry less often in general but I can't find it (if lockfile didn't change and everything is cached locally, why check?).

**Comment by weihanglo** [comment]
> cargo publish -p publish-disabled will error out, rather than skip the package. If we automatically skip packages, we'd need to decide in what circumstances we skip vs error and why.

True. Sorry I didn't think that hard. Also need to consider consistency between `cargo package` and `cargo publish`, where the former may not really care about `package.publish`

**Comment by epage** [comment]
I just ran into a failure doing a dry-run publish that I'm still trying to characterize.

This is on https://github.com/crate-ci/typos

```console
$ cargo +nightly publish --workspace --dry-run --exclude codespell-dict --exclude wikipedia-dict -Zpackage-workspace
...
   Verifying typos-dict v0.12.1 (/home/epage/src/personal/typos/crates/typos-dict)
error: failed to verify package tarball

Caused by:
  checksum for `varcon v1.0.1` changed between lock files

  this could be indicative of a few possible errors:

      * the lock file is corrupt
      * a replacement source in use (e.g., a mirror) returned a different checksum
      * the source itself may be corrupt in one way or another

  unable to verify that `varcon v1.0.1` is the same as when the lockfile was generated
```

This reproduces when I drop down to  `cargo package`
```console
$ cargo +nightly package --workspace -Zpackage-workspace
```

I thought it was because I had a local edit but that isn't true.  I just checked and `varcon` does not have any changes since the last release.

If I do these dry-runs on other repos, like `clap`, I don't have any problems like this.

**UPDATE**

the packaged `typos-dict` lockfile has
```toml
[[package]]
name = "varcon"
version = "1.0.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e9146d8774c298fff5715399e91d8495fe262198d3532b529cb73787b284b89d"
dependencies = [
 "varcon-core",
]
```
Creating a lockfile independent of any of this, I get
```toml
[[package]]
name = "varcon"
version = "1.0.1"
source = "registry+https://github.com/rust-lang/crates.io-index"
checksum = "e9146d8774c298fff5715399e91d8495fe262198d3532b529cb73787b284b89d"
dependencies = [
 "varcon-core",
]
```

<details>

<summary>Resolved</summary>

**UPDATE**

I looked closer at `clap` and `clap_complete` has `clap=4.5.22` in the lockfile, despite packaging `clap=4.5.23`.  I tried forcing the version req to `4.5.23` but I'm getting other errors
```
error: failed to prepare local package for uploading

Caused by:
  failed to select a version for `clap_lex`.
      ... required by package `clap_builder v4.5.23`
      ... which satisfies dependency `clap_builder = "=4.5.23"` of package `clap v4.5.23`
      ... which satisfies dependency `clap = "^4.5.23"` of package `clap_complete v4.5.40 (/home/epage/src/
personal/clap/clap_complete)`
  versions that meet the requirements `^0.7.4` are: 0.7.4

  all possible versions conflict with previously selected packages.

    previously selected package `clap_lex v0.7.1`
      ... which satisfies dependency `clap_lex = "^0.7.0"` (locked to 0.7.1) of package `clap_complete v4.5
.40 (/home/epage/src/personal/clap/clap_complete)`

  failed to select a version for `clap_lex` which could resolve this conflict
```

**UPDATE**

I'm thinking that the `clap_lex` problem is what is causing an older `clap` to be used.

My guess as to what is happening is that we are using the committed `Cargo.lock` as a starting point.  See #15622

</details

**Comment by kpreid** [comment]
> Might be good to open an issue for this.

Done: https://github.com/rust-lang/cargo/issues/15005

> If we automatically skip packages, we'd need to decide in what circumstances we skip vs error _and why_.

I don't think it _has be_ automatic, and to minimize the chance of proceeding with a user error, it shouldn’t be. Instead — actually, this is a nontrivial design space and so I filed a separate issue for it, https://github.com/rust-lang/cargo/issues/15006.

**Comment by epage** [comment]
I can't build that package.  I've even tried installing the system library that I think is needed

```
$ cargo check
   Compiling usbredirparser-sys v0.4.0 (/home/epage/src/personal/dump/usbredir-rs/usbredirparser-sys)
error: failed to run custom build command for `usbredirparser-sys v0.4.0 (/home/epage/src/personal/dump/usbredir-rs/us
bredirparser-sys)`

Caused by:
  process didn't exit successfully: `/home/epage/src/personal/dump/usbredir-rs/target/debug/build/usbredirparser-sys-7
2f98dc8c88e7a6b/build-script-build` (exit status: 101)
  --- stdout
  cargo:rerun-if-changed=/home/epage/src/personal/dump/usbredir-rs/usbredirparser-sys/Cargo.toml
  cargo:rerun-if-env-changed=LIBUSBREDIRPARSER_0.5_NO_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_PATH
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_LIBDIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64-unknown-linux-gnu
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR_x86_64_unknown_linux_gnu
  cargo:rerun-if-env-changed=HOST_PKG_CONFIG_SYSROOT_DIR
  cargo:rerun-if-env-changed=PKG_CONFIG_SYSROOT_DIR

  --- stderr
  thread 'main' panicked at usbredirparser-sys/build.rs:2:40:
  called `Result::unwrap()` on an `Err` value: PkgConfig(
  pkg-config exited with status code 1
  > PKG_CONFIG_ALLOW_SYSTEM_CFLAGS=1 pkg-config --libs --cflags libusbredirparser-0.5 libusbredirparser-0.5 >= 0.13

  The system library `libusbredirparser-0.5` required by crate `usbredirparser-sys` was not found.
  The file `libusbredirparser-0.5.pc` needs to be installed and the PKG_CONFIG_PATH environment variable must contain
its parent directory.
  The PKG_CONFIG_PATH environment variable is not set.

  HINT: if you have installed the library, try setting PKG_CONFIG_PATH to the directory containing `libusbredirparser-
0.5.pc`.
  )
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

```

**Comment by epage** [comment]
Still can't reproduce.  If you can create a repo that mimics the same behavior, independent of the other functionality, that would be great!

**Comment by fpoli** [comment]
@epage 
>     * Publish order is correct?
> 
>     * Handling of cycles?
> 
>     * Correctly waiting between published packages?
> 
>     * `--dry-run` works?

I just tried `cargo +nightly publish -Zpackage-workspace --workspace` with `cargo +nightly --version` = `cargo 1.88.0-nightly (0e93c5bf6 2025-04-05)`.

It seems that the command doesn't take dev-dependencies into account when computing the publish order. In my case, it fails to publish a workspace package X with dev-dependency on workspace package Y because Y has not been published yet. `cargo-release` has this exact same issue. See https://github.com/crate-ci/cargo-release/issues/829#issuecomment-2400217097.

**Comment by epage** [comment]
@fpoli yes, we should probably make that conditioned on whether the dependency will be removed.  Could you open an issue for that?

**Comment by epage** [comment]
I just ran into a non-dry-run failure.

I have a workspace somewhat like (very much simplified)
```
clap_complete_nushell (local) -> clap (local)
clap_complete_nushell (local) -> nu (dev)

nu -> clap
```

So when I went to publish `clap_complete_nushell`, the lockfile kept the registry `clap` dependency which was below the version requirement `clap_complete` had on `clap`.  To publish this, I had to run `cargo update clap`.  This means I would be unlikely to be able to release `clap` and `clap_complete_nushell` at the same time.

**Comment by epage** [comment]
In creating a test case and opening #15622, I found it is unrelated to this feature and can be reproduced without it.

**Comment by U007D** [comment]
This CfT has been added to TWiR Issue 602.

You may now remove the call-for-testing label. Please feel free to re-add the label if you wish this CfT to appear again in a future issue.

**Comment by epage** [comment]
Finally got a reproduction case for the "checksum changed" issue: #15647
